### PR TITLE
Added support for path param repeated fields

### PR DESCRIPTION
**Pull request by maros7**
Related to https://github.com/grpc-ecosystem/grpc-gateway/issues/741
Now it is possible to have repeated fields in the request, e.g. a message with `repeated string id = 1;` and  a RPC service `/orders/{id}` will make it possible to do a HTTP call like `GET /orders/<id>,<id>,...`. By default the plugin will generate code that expects separation by comma. But I also added support via a command line parameter to specify other separators supported as per OpenAPI 2.0. I.e. `csv`, `ssv`, `pipes` and `tsv`. Please note that the swagger-codegen doesn't generate a proper go client. I noted this in the issue I created. But the javascript client, used in the browser tests, works as expected.

### CODE REVIEW

**Comment by maros7** [comment]
All Travis jobs succeded except one with a failure not related to this PR. It says `HTTP request sent, awaiting response... 403 Forbidden` when fetching swagger-codegen.

File: protoc-gen-grpc-gateway/descriptor/registry.go [hunk 1]
```diff
@@ -50,12 +50,20 @@ type Registry struct {
...
+
 	// useJSONNamesForFields if true json tag name is used for generating fields in swagger definitions,
 	// otherwise the original proto name is used. It's helpful for synchronizing the swagger definition
 	// with grpc-gateway response, if it uses json tags for marshaling.
 	useJSONNamesForFields bool
 }
 
+type repeatedFieldSeparator struct {
+	name string
+	sep  string
```
Comment (johanbrandhorst) [inline protoc-gen-grpc-gateway/descriptor/registry.go]: If this is always a single character, should we just make it a `rune`?

File: protoc-gen-grpc-gateway/descriptor/types.go [hunk 2]
```diff
@@ -388,6 +428,28 @@ var (
...
+		descriptor.FieldDescriptorProto_TYPE_UINT64:  "runtime.Uint64Slice",
+		descriptor.FieldDescriptorProto_TYPE_INT32:   "runtime.Int32Slice",
+		descriptor.FieldDescriptorProto_TYPE_FIXED64: "runtime.Uint64Slice",
+		descriptor.FieldDescriptorProto_TYPE_FIXED32: "runtime.Uint32Slice",
+		descriptor.FieldDescriptorProto_TYPE_BOOL:    "runtime.BoolSlice",
+		descriptor.FieldDescriptorProto_TYPE_STRING:  "runtime.StringSlice",
+		// FieldDescriptorProto_TYPE_GROUP
+		// FieldDescriptorProto_TYPE_MESSAGE
+		// FieldDescriptorProto_TYPE_BYTES
+		// TODO(yugui) Handle bytes
```
Comment (johanbrandhorst) [inline protoc-gen-grpc-gateway/descriptor/types.go]: I suppose this is `TODO(maros7)` ;)?

File: protoc-gen-swagger/main.go [hunk 3]
```diff
@@ -15,13 +15,15 @@ import (
...
+	importPrefix               = flag.String("import_prefix", "", "prefix to be added to go package paths for imported proto files")
+	file                       = flag.String("file", "-", "where to load data from")
+	allowDeleteBody            = flag.Bool("allow_delete_body", false, "unless set, HTTP DELETE methods may not have a body")
+	grpcAPIConfiguration       = flag.String("grpc_api_configuration", "", "path to gRPC API Configuration in YAML format")
+	allowMerge                 = flag.Bool("allow_merge", false, "if set, generation one swagger file out of multiple protos")
+	mergeFileName              = flag.String("merge_file_name", "apidocs", "target swagger file name prefix after merge")
+	useJSONNamesForFields      = flag.Bool("json_names_for_fields", false, "if it sets Field.GetJsonName() will be used for generating swagger definitions, otherwise Field.GetName() will be used")
+	allowRepeatedFieldsInBody  = flag.Bool("allow_repeated_fields_in_body", false, "allows to use repeated field in `body` and `response_body` field of `google.api.http` annotation option")
```
Comment (johanbrandhorst) [inline protoc-gen-swagger/main.go]: Does this parameter make sense in the swagger generator? I feel like many of these don't make sense in this file unless I misunderstand what it's for :thinking:. This just generates the `x.swagger.json`, right? Why should it be allowed to error if there's a repeated field in the body or a body on a DELETE method?

Comment (johanbrandhorst) [reply protoc-gen-swagger/main.go]: (This isn't just directed at you, obviously, I can see the DELETE body flag was there already).

File: runtime/convert.go [hunk 4]
```diff
@@ -16,16 +17,57 @@ func String(val string) (string, error) {
...
 
+// StringSlice converts 'val' where individual strings are separated by
+// 'sep' into a string slice.
+func StringSlice(val, sep string) ([]string, error) {
+	s := strings.Split(val, sep)
+	values := make([]string, len(s))
+	for i, v := range s {
+		values[i] = v
+	}
+	return values, nil
```
Comment (johanbrandhorst) [inline runtime/convert.go]: Why are you creating a new slice? Just `return strings.Split(val, sep), nil`

**Comment by johanbrandhorst** [comment]
I reran the jobs as well and there seems to be one error: https://travis-ci.org/grpc-ecosystem/grpc-gateway/jobs/422808821

File: protoc-gen-grpc-gateway/descriptor/registry.go (hunk 1 above)
Comment (maros7) [reply protoc-gen-grpc-gateway/descriptor/registry.go]: You are quite right. I will change this.

File: protoc-gen-grpc-gateway/descriptor/types.go (hunk 2 above)
Comment (maros7) [reply protoc-gen-grpc-gateway/descriptor/types.go]: Of course ;) Question; any particular reason why bytes isn't implemented in proto2? Only been working with proto3 myself.

Comment (johanbrandhorst) [reply protoc-gen-grpc-gateway/descriptor/types.go]: I don't really know to be quite honest :grimacing:.

File: protoc-gen-swagger/main.go (hunk 3 above)
Comment (maros7) [reply protoc-gen-swagger/main.go]: Agreed, this was before the change `isPathParam` in [services.go](https://github.com/maros7/grpc-gateway/blob/repeated-field-path-parameter-support/protoc-gen-grpc-gateway/descriptor/services.go#L271). Will revert.

File: runtime/convert.go (hunk 4 above)
Comment (maros7) [reply runtime/convert.go]: Good point.. copy pasta of the other slice conversion functions. Will change.

**Comment by maros7** [comment]
Hmm, ok. Travis ran successfully on first commit, except the swagger codegen download thingy. Then I just added a missing test. But I will look into it.

**Comment by maros7** [comment]
I removed bin/ and vendor/, then re-ran tests (which also re-generates examples and runs dep/go install protoc) with make -B test. No errors here.

**Comment by maros7** [comment]
Travis ran with same result as master. I.e. one job ended with the same, unrelated failure:

```bash
0.32s$ test "${USE_BAZEL}" = true || sh -c 'cd examples/browser && node ./node_modules/gulp/bin/gulp'
module.js:538
    throw err;
    ^
Error: Cannot find module 'bower-logger'
    at Function.Module._resolveFilename (module.js:536:15)
    at Function.Module._load (module.js:466:25)
    at Module.require (module.js:579:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/home/travis/gopath/src/github.com/grpc-ecosystem/grpc-gateway/examples/browser/node_modules/bower/lib/commands/index.js:2:14)
    at Module._compile (module.js:635:30)
    at Object.Module._extensions..js (module.js:646:10)
    at Module.load (module.js:554:32)
    at tryModuleLoad (module.js:497:12)
    at Function.Module._load (module.js:489:3)
The command "test "${USE_BAZEL}" = true || sh -c 'cd examples/browser && node ./node_modules/gulp/bin/gulp'" exited with 1.
```

**Comment by johanbrandhorst** [comment]
Master has been fixed, please rebase and let's hope the tests pass.

**Comment by maros7** [comment]
Rebased and squashed. All tests pass now. Thanks!

**Comment by johanbrandhorst** [comment]
Thanks for your contribution!

**Pull request merged**